ssv/pre_consensus_justification.go

```diff
@@ -0,0 +1,148 @@
+package ssv
+
+import (
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	"github.com/bloxapp/ssv-spec/qbft"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/pkg/errors"
+)
+
+// correctQBFTState returns true if QBFT controller state requires pre-consensus justification
+func (b *BaseRunner) correctQBFTState(msg *qbft.SignedMessage) bool {
+	inst := b.QBFTController.InstanceForHeight(b.QBFTController.Height)
+	decidedInstance := inst != nil && inst.State != nil && inst.State.Decided
+
+	// firstHeightNotDecided is true if height == 0 (special case) and did not start yet
+	firstHeightNotDecided := inst == nil && b.QBFTController.Height == msg.Message.Height && msg.Message.Height == qbft.FirstHeight
+
+	// notFirstHeightDecided returns true if height != 0, height decided and the message is for next height
+	notFirstHeightDecided := decidedInstance && msg.Message.Height > qbft.FirstHeight && b.QBFTController.Height+1 == msg.Message.Height
+
+	return firstHeightNotDecided || notFirstHeightDecided
+}
+
+func (b *BaseRunner) shouldProcessingJustificationsForHeight(msg *qbft.SignedMessage) bool {
+	correctMsgTYpe := msg.Message.MsgType == qbft.ProposalMsgType || msg.Message.MsgType == qbft.RoundChangeMsgType
+	correctBeaconRole := b.BeaconRoleType == types.BNRoleProposer || b.BeaconRoleType == types.BNRoleAggregator || b.BeaconRoleType == types.BNRoleSyncCommitteeContribution
+	return b.correctQBFTState(msg) && correctMsgTYpe && correctBeaconRole
+}
+
+func (b *BaseRunner) validatePreConsensusJustifications(data *types.ConsensusData, highestDecidedDutySlot phase0.Slot) error {
+	//test invalid consensus data
+	if err := data.Validate(); err != nil {
+		return err
+	}
+
+	if b.BeaconRoleType != data.Duty.Type {
+		return errors.New("wrong beacon role")
+	}
+
+	if highestDecidedDutySlot >= data.Duty.Slot {
+		return errors.New("duty.slot < highest decided slot")
+	}
+
+	// validate justification quorum
+	if !b.Share.HasQuorum(len(data.PreConsensusJustifications)) {
+		return errors.New("no quorum")
+	}
+
+	signers := make(map[types.OperatorID]bool)
+	roots := make(map[[32]byte]bool)
+	rootCount := 0
+	for i, msg := range data.PreConsensusJustifications {
+		if err := msg.Validate(); err != nil {
+			return err
+		}
+
+		// check unique signers
+		if !signers[msg.Signer] {
```

Comment:
maybe do this before validate, hence we don't spend  work`Validate` for duplicate signers.

Reply:
but of course insert to the map only after validate. anyway we return err if we can't validate so even if we insert it's not that bad.

Reply:
can be optimized later